File: static/js/Vector_Operation.js
// =============================================
// OPERACIONES CON VECTORES
// =============================================

// ===============================
// Funciones de UI y manejo de inputs
// ===============================

/**
 * Limpia un campo de vector
 */
function clearVector(fieldId) {
    document.getElementById(fieldId).value = "";
}

// ===============================
// Operaciones con vectores y resultados
// ===============================

/**
 * Realiza una operación vectorial
 */
async function vectorOperation(operation) {
    try {
        // Determinar qué vector usar para operaciones individuales
        let selectedVector = document.querySelector('input[name="selectedVector"]:checked').value;

        const vector1 = document.getElementById("vector1").value;
        const vector2 = document.getElementById("vector2").value;

        let requestData = {
            operation: operation,
            vector1: selectedVector === "1" ? vector1 : vector2
        };

        // Operaciones que requieren dos vectores
        if (['add', 'subtract', 'dot', 'cross', 'angle', 'projection'].includes(operation)) {
            if (!vector1 || !vector2) {
                displayVectorResult("Error: Se necesitan ambos vectores para esta operación", "error");
                return;
            }
            requestData.vector1 = vector1;
            requestData.vector2 = vector2;
        } else {
            // Para operaciones individuales, verificar que el vector seleccionado exista
            if ((selectedVector === "1" && !vector1) || (selectedVector === "2" && !vector2)) {
                displayVectorResult(`Error: El vector ${selectedVector} está vacío`, "error");
                return;
            }
        }

        // Mostrar carga
        displayVectorResult("Calculando...", "loading");

        // Enviar al backend
        const response = await fetch('/vector_operation', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(requestData)
        });

        const data = await response.json();

        if (data.success) {
            displayVectorResult(data, "success");
        } else {
            displayVectorResult(`Error: ${data.error}`, "error");
        }

    } catch (error) {
        displayVectorResult(`Error: ${error.message}`, "error");
    }
}

/**
 * Muestra el resultado de una operación vectorial
 */
function displayVectorResult(data, type) {
    const resultDiv = document.getElementById("vector-result");

    if (type === "loading") {
        resultDiv.innerHTML = `<div class="text-center p-4">⏳ ${data}</div>`;
        return;
    }

    if (type === "error") {
        resultDiv.innerHTML = `
            <div class="text-center p-4 text-red-500">
                <p class="font-bold">Error</p>
                <p>${data}</p>
            </div>
        `;
        return;
    }

    if (type === "success") {
        let resultContent = `
            <div class="p-4">
                <div class="font-bold mb-2 text-lg text-center text-indigo-600">
                    ${getVectorOperationName(data.operation)}
                </div>
        `;

        if (Array.isArray(data.result)) {
            resultContent += `
                <div class="text-center my-4">
                    <div class="math-result text-xl mb-2">$$${data.latex_result}$$</div>
                    <div class="mt-3 p-3 bg-white rounded-lg shadow-inner border border-gray-200">
                        <p class="text-sm font-medium text-gray-600 mb-1">Forma vectorial:</p>
                        <p class="font-mono text-lg">[${data.result.join(', ')}]</p>
                    </div>
                </div>
            `;
        } else {
            resultContent += `
                <div class="text-center my-4">
                    <div class="math-result text-xl mb-3">$$${data.latex_result}$$</div>
                    <div class="mt-3 p-3 bg-white rounded-lg shadow-inner border border-gray-200">
                        <p class="text-sm font-medium text-gray-600 mb-1">Resultado:</p>
                        <p class="text-2xl font-bold">${data.result}</p>
                    </div>
                </div>
            `;
        }

        resultContent += `</div>`;
        resultDiv.innerHTML = resultContent;

        // Renderizar LaTeX con MathJax
        if (typeof MathJax !== 'undefined') {
            MathJax.typeset();
        }
    }
}

/**
 * Obtiene el nombre legible de una operación vectorial
 */
function getVectorOperationName(operation) {
    const names = {
        'add': 'Suma de Vectores',
        'subtract': 'Resta de Vectores',
        'dot': 'Producto Punto',
        'cross': 'Producto Cruz',
        'magnitude': 'Magnitud del Vector',
        'angle': 'Ángulo entre Vectores',
        'normalize': 'Vector Normalizado',
        'projection': 'Proyección Vectorial'
    };
    return names[operation] || operation;
}

File: routes/Vector_Operation.py
from flask import Blueprint, request, jsonify
import math
import sympy as sp

vector_bp = Blueprint('vector_bp', __name__)


# ********************************************
# *         OPERACIONES CON VECTORES         *
# ********************************************

@vector_bp.route('/vector_operation', methods=['POST'])
def vector_operation():
    data = request.json
    operation = data['operation']
    
    try:
        # Parsear los vectores
        vector1 = [float(x.strip()) for x in data['vector1'].split(',')]
        vector2 = [float(x.strip()) for x in data['vector2'].split(',')] if 'vector2' in data else None
        
        result = None
        details = ""
        latex_result = ""
        
        if operation == 'add':
            if len(vector1) != len(vector2):
                raise ValueError("Los vectores deben tener la misma dimensión")
            result = [a + b for a, b in zip(vector1, vector2)]
            details = f"Suma de vectores"
            latex_result = sp.latex(sp.Matrix(vector1)) + " + " + sp.latex(sp.Matrix(vector2)) + " = " + sp.latex(sp.Matrix(result))
            
        elif operation == 'subtract':
            if len(vector1) != len(vector2):
                raise ValueError("Los vectores deben tener la misma dimensión")
            result = [a - b for a, b in zip(vector1, vector2)]
            details = f"Resta de vectores"
            latex_result = sp.latex(sp.Matrix(vector1)) + " - " + sp.latex(sp.Matrix(vector2)) + " = " + sp.latex(sp.Matrix(result))
            
        elif operation == 'dot':
            if len(vector1) != len(vector2):
                raise ValueError("Los vectores deben tener la misma dimensión")
            result = sum(a * b for a, b in zip(vector1, vector2))
            details = f"Producto punto"
            latex_result = sp.latex(sp.Matrix(vector1)) + " \\cdot " + sp.latex(sp.Matrix(vector2)) + " = " + sp.latex(result)
            
        elif operation == 'cross':
            if len(vector1) != 3 or len(vector2) != 3:
                raise ValueError("El producto cruz solo está definido para vectores 3D")
            result = [
                vector1[1]*vector2[2] - vector1[2]*vector2[1],
                vector1[2]*vector2[0] - vector1[0]*vector2[2],
                vector1[0]*vector2[1] - vector1[1]*vector2[0]
            ]
            details = f"Producto cruz"
            latex_result = sp.latex(sp.Matrix(vector1)) + " \\times " + sp.latex(sp.Matrix(vector2)) + " = " + sp.latex(sp.Matrix(result))
            
        elif operation == 'magnitude':
            result = math.sqrt(sum(x**2 for x in vector1))
            details = f"Magnitud del vector"
            latex_result = "\\|" + sp.latex(sp.Matrix(vector1)) + "\\| = " + sp.latex(result)
            
        elif operation == 'angle':
            if len(vector1) != len(vector2):
                raise ValueError("Los vectores deben tener la misma dimensión")
            dot_product = sum(a * b for a, b in zip(vector1, vector2))
            mag1 = math.sqrt(sum(x**2 for x in vector1))
            mag2 = math.sqrt(sum(x**2 for x in vector2))
            angle_rad = math.acos(dot_product / (mag1 * mag2))
            result = math.degrees(angle_rad)
            details = f"Ángulo entre vectores (grados)"
            latex_result = "\\theta = \\arccos\\left(\\frac{" + sp.latex(sp.Matrix(vector1)) + " \\cdot " + sp.latex(sp.Matrix(vector2)) + "}{\\|" + sp.latex(sp.Matrix(vector1)) + "\\| \\cdot \\|" + sp.latex(sp.Matrix(vector2)) + "\\|}\\right) = " + sp.latex(result) + "^\\circ"
            
        elif operation == 'normalize':
            magnitude = math.sqrt(sum(x**2 for x in vector1))
            if magnitude == 0:
                raise ValueError("No se puede normalizar el vector cero")
            result = [x/magnitude for x in vector1]
            details = f"Vector normalizado"
            latex_result = "\\hat{" + sp.latex(sp.Matrix(vector1)) + "} = \\frac{" + sp.latex(sp.Matrix(vector1)) + "}{\\|" + sp.latex(sp.Matrix(vector1)) + "\\|} = " + sp.latex(sp.Matrix(result))
            
        elif operation == 'projection':
            if len(vector1) != len(vector2):
                raise ValueError("Los vectores deben tener la misma dimensión")
            dot_product = sum(a * b for a, b in zip(vector1, vector2))
            mag2_squared = sum(x**2 for x in vector2)
            if mag2_squared == 0:
                raise ValueError("No se puede proyectar sobre el vector cero")
            scalar = dot_product / mag2_squared
            result = [x * scalar for x in vector2]
            details = f"Proyección de v1 sobre v2"
            latex_result = "\\text{proj}_{" + sp.latex(sp.Matrix(vector2)) + "}" + sp.latex(sp.Matrix(vector1)) + " = \\left(\\frac{" + sp.latex(sp.Matrix(vector1)) + " \\cdot " + sp.latex(sp.Matrix(vector2)) + "}{\\|" + sp.latex(sp.Matrix(vector2)) + "\\|^2}\\right)" + sp.latex(sp.Matrix(vector2)) + " = " + sp.latex(sp.Matrix(result))
            
        return jsonify({
            'success': True,
            'result': result,
            'latex_result': latex_result,
            'details': details,
            'operation': operation
        })
        
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e),
            'operation': operation
        })